minifi-nar-bundles/minifi-provenance-reporting-bundle/minifi-provenance-reporting-task/src/main/java/org/apache/nifi/minifi/provenance/reporting/ProvenanceReportingTask.java

```diff
@@ -0,0 +1,443 @@
+                return;
+            }
+            if (state.containsKey(LAST_EVENT_ID_KEY)) {
+                firstEventId = Long.parseLong(state.get(LAST_EVENT_ID_KEY)) + 1;
+            }
+        }
+
+        final List<ProvenanceEventRecord> events;
+        try {
+            events = context.getEventAccess().getProvenanceEvents(firstEventId, context.getProperty(BATCH_SIZE).asInteger());
+        } catch (final IOException ioe) {
+            getLogger().error("Failed to retrieve Provenance Events from repository due to {}", ioe);
+            return;
+        }
+
+        if (events == null || events.isEmpty()) {
+            getLogger().debug("No events to send.");
+            return;
+        }
+
+        final long start = System.nanoTime();
+        final Map<String, ?> config = Collections.emptyMap();
+        final JsonBuilderFactory factory = Json.createBuilderFactory(config);
+        final JsonObjectBuilder builder = factory.createObjectBuilder();
+
+        final String nifiUrl = context.getProperty(MINIFI_URL).evaluateAttributeExpressions().getValue();
+        URL url;
+        try {
+            url = new URL(nifiUrl);
+        } catch (final MalformedURLException e1) {
+            // already validated
+            throw new AssertionError();
+        }
+
+        final String hostname = url.getHost();
+
+        final JsonArrayBuilder arrayBuilder = factory.createArrayBuilder();
+        for (final ProvenanceEventRecord event : events) {
+            arrayBuilder.add(serialize(factory, builder, event, getComponentName(procGroupStatus, event), hostname, url, rootGroupName));
+        }
+        final JsonArray jsonArray = arrayBuilder.build();
+
+        try {
+            final Transaction transaction = getClient().createTransaction(TransferDirection.SEND);
+            if (transaction == null) {
+                getLogger().debug("All destination nodes are penalized; will attempt to send data later");
+                return;
+            }
+
+            final Map<String, String> attributes = new HashMap<>();
+            final String transactionId = UUID.randomUUID().toString();
+            attributes.put("reporting.task.transaction.id", transactionId);
+
+            final byte[] data = jsonArray.toString().getBytes(StandardCharsets.UTF_8);
+            transaction.send(data, attributes);
+            transaction.confirm();
+            transaction.complete();
+
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+            getLogger().info("Successfully sent {} Provenance Events to destination in {} ms; Transaction ID = {}; First Event ID = {}",
+                new Object[] {events.size(), transferMillis, transactionId, events.get(0).getEventId()});
+        } catch (final IOException e) {
+            getLogger().error("Failed to send Provenance Events to destination due to {}", new Object[] {e}, e);
```

Comment:
I think some handling of the exception is needed, otherwise if there is an exception indices would likely be off with the following bit of code in terms of taking care of items even if nothing is sent.  In context of below, I think we should opt for at least once semantics.